File: topos/mcp/tools/docs.py
"""
Documentation tool — read the Topos docs as an MCP tool.

Exists because MCP resources (``topos://docs/*``) are not uniformly reachable
from agents across clients in 2026. Claude Code bridges resources to the
model implicitly; Gemini CLI (and others) do not. This tool exposes the same
markdown content via a tool call so critical context (especially the
workflow guide) is reachable everywhere.

Content is read verbatim from ``../resources/content/*.md`` — single source
of truth shared with the resource handlers.
"""

from __future__ import annotations

from pathlib import Path
from typing import Literal

from ..server import mcp

_CONTENT_DIR = Path(__file__).parent.parent / "resources" / "content"

DocTopic = Literal[
    "agent-contract", "lattice", "metrics", "preferences", "priority", "workflows"
]


@mcp.tool(
    name="topos_get_doc",
    tags={"docs"},
    annotations={
        "title": "Topos Documentation",
        "readOnlyHint": True,
        "destructiveHint": False,
        "idempotentHint": True,
        "openWorldHint": False,
    },
)
def topos_get_doc(topic: DocTopic) -> str:
    """Return a Topos documentation page as Markdown.

    Use when your MCP client does not expose resource fetching to the agent
    (e.g. Gemini CLI). Clients that do surface resources should prefer the
    equivalent resource URI for efficiency: ``topos://docs/{topic}``.

    Topics:
        agent-contract — compact outcome-first loop contract and done gates.
        lattice    — the 8-element 3-cube H(G_qual); top = IDEAL, bottom = SLOP.
        metrics    — every metric key, thresholds, interpretation.
        preferences — strict generator rankings and preference walks.
        priority   — priority profiles (simple/composable/secure).
        workflows  — the canonical review→plan→refactor→re-measure loop.
                     **Read this first on every new refactor session.**
    """
    path = _CONTENT_DIR / f"{topic}.md"
    return path.read_text(encoding="utf-8")

File: topos/mcp/src/docs.rs
//! Embedded documentation content, shared by the `topos_get_doc` tool and
//! the `topos://docs/*` resources.
//!
//! Content is compiled in via `include_str!` (from `docs/content/*.md`),
//! so the server is a single self-contained binary — no runtime file reads,
//! matching the "single source of truth" the Python package achieved with
//! a shared content directory.

use crate::schemas::DocTopic;

pub const AGENT_CONTRACT: &str = include_str!("../docs/content/agent-contract.md");
pub const LATTICE: &str = include_str!("../docs/content/lattice.md");
pub const METRICS: &str = include_str!("../docs/content/metrics.md");
pub const PREFERENCES: &str = include_str!("../docs/content/preferences.md");
pub const PRIORITY: &str = include_str!("../docs/content/priority.md");
pub const WORKFLOWS: &str = include_str!("../docs/content/workflows.md");

/// Content for a documentation topic.
pub fn doc_content(topic: DocTopic) -> &'static str {
    match topic {
        DocTopic::AgentContract => AGENT_CONTRACT,
        DocTopic::Lattice => LATTICE,
        DocTopic::Metrics => METRICS,
        DocTopic::Preferences => PREFERENCES,
        DocTopic::Priority => PRIORITY,
        DocTopic::Workflows => WORKFLOWS,
    }
}

/// Content for a `topos://docs/<slug>` resource URI, or `None` for an
/// unknown slug.
pub fn doc_content_for_slug(slug: &str) -> Option<&'static str> {
    match slug {
        "agent-contract" => Some(AGENT_CONTRACT),
        "lattice" => Some(LATTICE),
        "metrics" => Some(METRICS),
        "preferences" => Some(PREFERENCES),
        "priority" => Some(PRIORITY),
        "workflows" => Some(WORKFLOWS),
        _ => None,
    }
}

/// The six resource slugs, in listing order.
pub const DOC_SLUGS: [&str; 6] = [
    "agent-contract",
    "lattice",
    "metrics",
    "priority",
    "preferences",
    "workflows",
];
